org.scala-ide.sdt.debug/src/org/scalaide/debug/internal/async/RetainedStackManager.scala

```diff
@@ -0,0 +1,135 @@
+    stackFrames += (body.asInstanceOf[ObjectReference], messageOrdinal) -> mkStackTrace(frames, messageOrdinal)
+  }
+
+  private def mkStackTrace(frames: Seq[StackFrame], ordinal: Int): AsyncStackTrace = {
+    import collection.JavaConverters._
+
+    val asyncFrames = frames.map { frame =>
+      Try {
+        val names = frame.visibleVariables()
+        val values = frame.getValues(names)
+        val locals = values.asScala.map {
+          case (lvar, lval) => AsyncLocalVariable(lvar.name(), ScalaValue(lval, debugTarget), ordinal)(debugTarget)
+        }
+        val location = frame.location()
+        AsyncStackFrame(locals.toSeq, Location(location.sourceName, location.declaringType.name, location.lineNumber))(debugTarget)
+      }
+    } collect {
+      case Success(asyncStackTrace) => asyncStackTrace
+    }
+
+    AsyncStackTrace(asyncFrames)
+  }
+
+  private def installMethodBreakpoint(tpe: ReferenceType, app: AsyncProgramPoint): Unit = {
+    val method = AsyncUtils.findAsyncProgramPoint(app, tpe)
+    method.foreach { meth =>
+      val req = JdiRequestFactory.createMethodEntryBreakpoint(method.get, debugTarget)
+      debugTarget.eventDispatcher.setActorFor(actor, req)
+      req.putProperty("app", app)
+      req.enable()
+
+      logger.debug(s"Installed method breakpoint for ${method.get.declaringType()}.${method.get.name}")
+    }
+  }
+
+  private val programPoints = List(
+    AsyncProgramPoint("scala.concurrent.Future$", "apply", 0),
+    AsyncProgramPoint("scala.concurrent.package$", "future", 0),
+    AsyncProgramPoint("play.api.libs.iteratee.Cont$", "apply", 0),
+    AsyncProgramPoint("akka.actor.LocalActorRef", "$bang", 0),
+    AsyncProgramPoint("akka.actor.RepointableActorRef", "$bang", 0),
+    AsyncProgramPoint("scala.actors.InternalReplyReactor$class", "$bang", 1))
+
+  /** Return the saved stackframes for the given future body (if any). */
+  def getStackFrameForFuture(future: ObjectReference, messageOrdinal: Int): Option[AsyncStackTrace] =
+    stackFrames.get(previousKeyWhenIsGiven(future, messageOrdinal))
+
+  private def keysForObjectReferenceAndOrdinal(future: ObjectReference, ordinal: Int): PartialFunction[AsyncStackTraceKey, AsyncStackTraceKey] = {
+    case k @ (objRef, ord) if objRef == future && ord < ordinal => k
+  }
+
+  private def keysForObjectReference(future: ObjectReference): PartialFunction[AsyncStackTraceKey, AsyncStackTraceKey] = {
+    case k @ (objRef, _) if objRef == future => k
+  }
+
+  private def keysSelection(future: ObjectReference): PartialFunction[Int, PartialFunction[AsyncStackTraceKey, AsyncStackTraceKey]] = {
+    case ordinal if ordinal != OrdinalNotSet => keysForObjectReferenceAndOrdinal(future, ordinal)
+    case OrdinalNotSet => keysForObjectReference(future)
+  }
+
+  private def previousKeyWhenIsGiven(future: ObjectReference, ordinal: Int): AsyncStackTraceKey = {
+    val olderTracesKeys = stackFrames.keys.collect(keysSelection(future)(ordinal))
```

Comment:
I don't like that, for the special-case of objects, we start doing a linear search for everybody, instead of a hashmap lookup. There should be a fast-path for objects that are unique. If there's only one entry for some `future` (we also need to rename this, because it's not a future, it can be a message as well), there's no need to look further. So maybe go back to a `Map[ObjectReference, <something else here>]`, and have a fast-path when there's only one of those, and otherwise use the `ordinal` to disambiguate.

Also, please add an explanation in the scaladoc comment of `getStackFrameForFuture` (also, let's rename it to reflect that it's not only for futures... maybe simply `getAsyncStackFrame`).